client/pkg/connectionctx/manager.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2024 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connectionctx
+
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/exp/rand"
+)
+
+type connectionCtx[T any] struct {
+	Ctx    context.Context
+	Cancel context.CancelFunc
+	// Current URL of the stream connection.
+	StreamURL string
+	// Current stream to send the gRPC requests.
+	Stream T
+}
+
+// Manager is used to manage the connection contexts.
+type Manager[T any] struct {
+	sync.RWMutex
+	connectionCtxs map[string]*connectionCtx[T]
+}
+
+// NewManager is used to create a new connection context manager.
+func NewManager[T any]() *Manager[T] {
+	return &Manager[T]{
+		connectionCtxs: make(map[string]*connectionCtx[T], 3),
+	}
+}
+
+// Exist is used to check if the connection context exists by the given URL.
+func (c *Manager[T]) Exist(url string) bool {
+	c.RLock()
+	defer c.RUnlock()
+	_, ok := c.connectionCtxs[url]
+	return ok
+}
+
+// Store is used to store the connection context, `overwrite` is used to force the store operation
+// no matter whether the connection context exists before, which is false by default.
+func (c *Manager[T]) Store(ctx context.Context, url string, stream T, overwrite ...bool) {
+	c.Lock()
+	defer c.Unlock()
+	overwriteFlag := false
+	if len(overwrite) > 0 {
+		overwriteFlag = overwrite[0]
+	}
+	_, ok := c.connectionCtxs[url]
+	if !overwriteFlag && ok {
+		return
+	}
+	c.storeLocked(ctx, url, stream)
+}
+
+func (c *Manager[T]) storeLocked(ctx context.Context, url string, stream T) {
+	c.releaseLocked(url)
+	cctx, cancel := context.WithCancel(ctx)
+	c.connectionCtxs[url] = &connectionCtx[T]{cctx, cancel, url, stream}
```

Comment:
Will it be overwritten every time?

Reply:
Currently, whether to overwrite the original `connectionCtx` is handled before `storeLocked` is checked, and `releaseLocked` will also clean the old one before overwriting.